include/revng/Yield/Internal/Instruction.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <limits>
+#include <string>
+
+#include "revng/ADT/SortedVector.h"
+#include "revng/Model/VerifyHelper.h"
+#include "revng/Support/MetaAddress.h"
+#include "revng/Support/MetaAddress/YAMLTraits.h"
+#include "revng/Yield/Internal/TagType.h"
+
+namespace yield {
+using ByteContainer = llvm::SmallVector<uint8_t, 16>;
+}
+
+/* TUPLE-TREE-YAML
+
+name: Instruction
+type: struct
+fields:
+  - name: Address
+    type: MetaAddress
+  - name: Bytes
+    type: yield::ByteContainer
+  - name: Raw
+    type: std::string
+
+  - name: Tags
+    sequence:
+      type: std::vector
+      elementType: yield::Tag
+    optional: true
+
+  - name: Opcode
+    type: std::string
+    optional: true
+  - name: Comment
+    type: std::string
+    optional: true
+  - name: Error
+    type: std::string
+    optional: true
+
+  - name: HasDelayedSlot
```

Comment:
What is the semantic of this field? Is this `yield::Instruction` representing both the jump instruction and the instruction in the delay slot?

```suggestion
  - name: HasDelaySlot
```

Reply:
This field is set to true for an instructions that contains a delayed slot. The actual delayed instruction is an instruction that follows the one with this field set.

I went with this approach because that's how MCDissassembler stores these. This allows generate instructions independently from each other - no need to carry over any information between neighbouring ones.

Reply:
OK, document this. Unless you saw *delayed* somewhere let's go with *delay*.

Also, in the verify assert that this is the semi-last instruction of the list.

Reply:
I looked it up, it's my bad. For some reason I mixed up `delayed load` and `delay slot` when naming this. Will fix.

2. Do we care where it is really? Sure, we should probably check that it's not the last, but other than that let the instruction be wherever it is, imo.

Reply:
1. :+1:
2. Mmh, I like our data structures to offer strong, useful, predictable guarantees. If in our mind we expect something, why not enforce it?

Reply:
2. Mostly, in this case, if we try to enforce something this rigid, I see no point of having this flag at an instruction level. If the delay slot can only be located at the very end of a block, it should be a block-level flag!

Now that I think about it, maybe that's actually the way to go. What do you think? Should `EndsInDelaySlot` be a basic-block field after all?

Reply:
Yes!

Reply:
What do you think?
https://github.com/revng/revng/pull/179/commits/c514425b189fe2fcbf337ff546726575e8c3c0f6

Reply:
Looks good!

* I'd go for `HasDelaySlot`, there's doc for better specifying the meaning.
* `InstructionReturnValueTuple`: you often do this, can you find a more semantic name instead of simply describing the role? I find that providing more dignity to data structure is always beneficial.

Reply:
1. Sure.
2. I do it for the data structures I wouldn't have added if I could get away with it. For example, in this case it's literally `std::tuple<yield::Instruction, bool, size_t>` except that's too long and somewhat confusing (What does this `bool` mean? What's the int for?), hence the struct with named fields instead. If you have any name suggestions - I'm not opposed, BUT, it's only ever used as the return value of this specific function and is expected to be unrolled using structured bindings at the call site.

Reply:
2. It wouldn't provide it much dignity to call it its real name: `InstructionAndTheFieldsThatWereRemovedFromIt` :0

Reply:
I'd call it "LLVMDisassemblerInterface::Instruction" and be done with it. Its scoped, no one will use it for other purpose, it makes sense as a name. Occam's razor!

Reply:
I guess that's one way to do it. It is somewhat confusing to have `Instruction` class that has an `Instruction` member with type `Instruction` but, either way, this is not much worse the the current solution, so I'm fine with that.